src/service_alerts.rs

```diff
@@ -23,40 +24,51 @@ pub const FINALIZER_NAME: &str = "servicealerter.cactuar.rs";
 )]
 pub struct ServiceAlerterSpec {
     pub common_labels: HashMap<String, String>,
-    pub service_selector: ServiceSelector,
-
-    // For reasons to do with `kube` relying on JsonSchema, I cannot use serde's
-    // 'with' attribute here, so you need to manually disambiguate both the
-    // serialisation and deserialisation.
-    //
-    // See: https://github.com/GREsau/schemars/issues/89
-    #[serde(
-        serialize_with = "serde_yaml::with::singleton_map_recursive::serialize",
-        deserialize_with = "serde_yaml::with::singleton_map_recursive::deserialize"
-    )]
-    pub alerts: Vec<Alert>,
+    pub deployment_name: String,
+    pub alerts: HashMap<Alerts, Vec<AlertConfig>>,
 }
 
-#[derive(Debug, Serialize, Deserialize, Clone, JsonSchema, PartialEq)]
+// #[derive(Debug, Serialize, Deserialize, Clone, JsonSchema, PartialEq)]
+// #[serde(rename_all = "camelCase")]
+// pub struct Alerts {
+//     replica_count: Option<Vec<AlertConfig>>,
+//     error_percent: Option<Vec<AlertConfig>>,
+//     traffic_per_second: Option<Vec<AlertConfig>>,
+//     latency_histogram_milliseconds: Option<Vec<AlertConfigHistogram>>,
+// }
+
+#[derive(Debug, Serialize, Deserialize, Clone, JsonSchema, PartialEq, Eq, Hash)]
 #[serde(rename_all = "camelCase")]
-pub struct ServiceSelector {
-    pub name: String,
+pub enum Alerts {
+    ReplicaCount,
+    ErrorPercent,
+    TrafficPerSecond,
+    LatencyMillisecondsP50,
+    LatencyMillisecondsP90,
+    LatencyMillisecondsP95,
+    LatencyMillisecondsP99,
 }
 
 #[derive(Debug, Serialize, Deserialize, Clone, JsonSchema, PartialEq)]
 #[serde(rename_all = "camelCase")]
-pub enum Alert {
-    Replicas(SeverityBoundary),
-    RequestErrorPercent(SeverityBoundary),
+pub struct AlertConfig {
+    operation: Operation,
+    value: f32,
+    #[serde(rename = "for")]
+    for_: String, // want to be able to specify like 3m 4s
+    alert_with_labels: HashMap<String, String>,
 }
 
-type SeverityBoundary = HashMap<Severity, Boundary>;
-
+// Kubernetes enums start with an upper case letter
 #[derive(Debug, Serialize, Deserialize, Clone, JsonSchema, Eq, PartialEq, Hash)]
 #[serde(rename_all = "camelCase")]
-pub enum Boundary {
-    LessThan(i32),
-    EqualTo(i32),
+pub enum Operation {
+    #[serde(rename = "EqualTo")]
+    EqualTo,
+    #[serde(rename = "LessThan")]
+    LessThan,
+    #[serde(rename = "MoreThan")]
+    MoreThan,
```

Comment:
You're overriding the `#serde(rename_all = "camelCase")]` macro here with the individual `rename` calls. By default, Serde should just replicate the names, so you can probably just drop all of the `renames`:
```rust
#[derive(Debug, Serialize, Deserialize, Clone, JsonSchema, Eq, PartialEq, Hash)]
pub enum Operation {
    EqualTo,
    LessThan,
    MoreThan,
}
```

Reply:
ah great, thanks will fix